### ISSUE TITLE
Claude Code Review Report

### ISSUE THREAD

**Issue by qarmin** [sahib/rmlint#771]
Recently, I ran Claude Code across several projects(like image-rs, lofty of symphonia) to look for potential issues(logic bugs, copy paste errors, architecture problems, panics etc.).

In this projects, a significant portion of reported findings turned out to be incorrect or minor false positives, but a non-trivial subset was valid and included real issues ranging from incorrect comments to actual logic bugs. As a result, the findings generally require manual validation to separate noise from actionable problems.

Full Report - [rmlint_20260619.html](https://github.com/user-attachments/files/29145486/rmlint_20260619.html)

Example findings (this is only a subset of the findings, specifically those most likely to be actual bugs; for the full list, see the report above):

=====================================

### CPY_1 `MEDIUM` ⚡ easy

**Description:** In rm_file_cmp_without_extension(), when computing the comparison length for file B's basename, the fallback (no-extension) branch incorrectly uses `strlen(basename_a)` instead of `strlen(basename_b)`: `size_t b_len = (ext_b) ? (ext_b - basename_b) : (int)strlen(basename_a);`. When file B has no extension and its basename length differs from file A's, b_len is wrong, which can cause incorrect SIGN_DIFF results and/or g_ascii_strncasecmp comparing the wrong number of characters under --match-without-extension. This can misclassify or fail to classify files as duplicates/originals based on basename matching.

**Locations:**

```
lib/preprocess.c:49-63
   49 | static gint rm_file_cmp_without_extension(const RmFile *file_a, const RmFile *file_b) {
   50 |     const char *basename_a = file_a->folder->basename;
   51 |     const char *basename_b = file_b->folder->basename;
   52 | 
   53 |     char *ext_a = rm_util_path_extension(basename_a);
   54 |     char *ext_b = rm_util_path_extension(basename_b);
   55 | 
   56 |     /* Check length till extension, or full length if none present */
   57 |     size_t a_len = (ext_a) ? (ext_a - basename_a) : (int)strlen(basename_a);
   58 |     size_t b_len = (ext_b) ? (ext_b - basename_b) : (int)strlen(basename_a);
   59 | 
   60 |     RETURN_IF_NONZERO(SIGN_DIFF(a_len, b_len));
   61 | 
   62 |     return g_ascii_strncasecmp(basename_a, basename_b, a_len);
   63 | }
```

**Fix:** Change line 58 to `size_t b_len = (ext_b) ? (ext_b - basename_b) : (int)strlen(basename_b);`.

### CPY_2 `LOW` ⚡ easy

**Description:** The inversed_option_entries[] table in rm_cmd_parse_args() contains the exact same GOptionEntry `{"no-match-extension", 'E', DISABLE | HIDDEN, G_OPTION_ARG_NONE, &cfg->match_with_extension, "Disable --match-extension", NULL}` listed twice in a row (lines 1477 and 1478), registering the long option "no-match-extension" and short option -E twice within the same GOptionGroup passed to g_option_group_add_entries(). This is clearly a copy-paste leftover (the very next, distinct entry at line 1479 for "no-match-without-extension"/-I suggests the second copy of line 1477 should likely not exist, or was meant to be something else entirely).

**Locations:**

```
lib/cmdline.c:1477-1478
 1477 |         {"no-match-extension"         , 'E' , DISABLE | HIDDEN , G_OPTION_ARG_NONE     , &cfg->match_with_extension    , "Disable --match-extension"           , NULL} ,
 1478 |         {"no-match-extension"         , 'E' , DISABLE | HIDDEN , G_OPTION_ARG_NONE     , &cfg->match_with_extension    , "Disable --match-extension"           , NULL} ,
```

**Fix:** Remove the duplicate line 1478 (or 1477); keep only a single registration of the "no-match-extension"/-E option.

### CMT_1 `LOW` ⚡ easy

**Description:** Runner.run()'s docstring states 'Returns: a `Script` instance.' but the method has no return statement at all (implicitly returns None) and never constructs a Script object anywhere in its body.

**Locations:**

```
gui/shredder/runner.py:327-330
  327 |     def run(self):
  328 |         """Trigger the run of the rmlint process.
  329 |         Returns: a `Script` instance.
  330 |         """
```

**Fix:** Remove the incorrect Returns clause from the docstring, or change run() to actually return something meaningful if that was the original intent.

### LOGIC_4 `HIGH` ⚡ easy

**Description:** rm_xattr_is_deduplicated() is declared to return bool where true means 'already deduplicated' (per its own doc comment and its only caller in rm_session_dedupe_main, which skips the dedupe work entirely when true is returned). On a failed rm_sys_stat() it does 'return EXIT_FAILURE;' which is the integer 1, i.e. true when implicitly converted to bool. So any stat() failure (e.g. permission denied, race with file removal) is reported as 'file IS already deduplicated', causing rm_session_dedupe_main to silently skip deduplication instead of treating the check as inconclusive/failed.

**Locations:**

```
lib/xattr.c:406-413
  406 | bool rm_xattr_is_deduplicated(const char *path, bool follow_symlinks) {
  407 |     g_assert(path);
  408 | 
  409 |     RmStat stat_buf;
  410 |     if(rm_sys_stat(path, &stat_buf) < 0) {
  411 |         rm_log_warning_line("failed to check dedupe state of %s: %s", path, g_strerror(errno));
  412 |         return EXIT_FAILURE;
  413 |     }
```

**Fix:** Return false (not EXIT_FAILURE) when the stat() call fails, since this function's contract uses bool/true=deduplicated, not a process exit code.

### MEM_3 `LOW` ⚡ easy

**Description:** rm_digest_copy() allocates `copy` via g_slice_copy(sizeof(RmDigest), digest) before checking whether interface->copy is NULL; if it is NULL, the function returns NULL directly without freeing the already-allocated `copy` slice, leaking sizeof(RmDigest) bytes. Currently unreachable since every registered RmDigestInterface in this file defines a non-NULL .copy function, but the early-return path is dead-but-buggy code that will leak if a future digest type is added without a copy implementation.

**Locations:**

```
lib/checksum.c:944-956
  944 | RmDigest *rm_digest_copy(RmDigest *digest) {
  945 |     g_assert(digest);
  946 | 
  947 |     RmDigest *copy = g_slice_copy(sizeof(RmDigest), digest);
  948 | 
  949 |     const RmDigestInterface *interface = rm_digest_get_interface(digest->type);
  950 |     if(interface->copy == NULL) {
  951 |         return NULL;
  952 |     }
  953 | 
  954 |     copy->state = interface->copy(digest->state);
  955 |     return copy;
  956 | }
```

**Fix:** Free `copy` with g_slice_free(RmDigest, copy) before returning NULL when interface->copy is NULL, or restructure to check interface->copy before allocating copy.

**Comment by vassilit** [comment]
CPY_1 is fixed on develop in 2d6c648
CPY_2 is fixed on develop in 16f2163

**Comment by vassilit** [comment]
DEAD_2 and DEAD_3 are fixed on develop in 279a278
OVF_5 and OVF_9 are fixed on develop in 0e61e2f5
